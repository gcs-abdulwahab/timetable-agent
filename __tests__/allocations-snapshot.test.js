/**
 * Allocations Snapshot Test
 * 
 * This test ensures that the allocations.json file structure and sort order
 * remain consistent to prevent unintentional changes to the data format.
 */

const fs = require('fs');
const path = require('path');

describe('Allocations Data Structure', () => {
  let allocationsData;
  
  beforeAll(() => {
    const allocationsPath = path.join(__dirname, '..', 'data', 'allocations.json');
    const content = fs.readFileSync(allocationsPath, 'utf8');
    // Remove BOM if present
    const cleanContent = content.replace(/^\uFEFF/, '');
    allocationsData = JSON.parse(cleanContent);
  });

  test('allocations.json structure matches expected format', () => {
    // Test that allocations is an array
    expect(Array.isArray(allocationsData)).toBe(true);
    
    // Test that we have some allocations
    expect(allocationsData.length).toBeGreaterThan(0);
    
    // Create a snapshot of the first few allocations to verify structure
    const sampleAllocations = allocationsData.slice(0, 3).map(allocation => {
      // Create a normalized version for testing structure
      return {
        id: typeof allocation.id,
        subjectId: typeof allocation.subjectId,
        teacherId: typeof allocation.teacherId,
        timeSlotId: typeof allocation.timeSlotId,
        day: typeof allocation.day,
        room: typeof allocation.room,
        semesterId: typeof allocation.semesterId,
        // Include actual values for key fields to detect changes
        hasId: !!allocation.id,
        hasSubjectId: !!allocation.subjectId,
        hasTimeSlotId: !!allocation.timeSlotId,
        hasDay: !!allocation.day,
        hasSemesterId: !!allocation.semesterId,
        allowedFields: Object.keys(allocation).sort()
      };
    });
    
    expect(sampleAllocations).toMatchSnapshot();
  });

  test('all allocations have required fields', () => {
    const requiredFields = ['id', 'subjectId', 'timeSlotId', 'day', 'semesterId'];
    
    allocationsData.forEach((allocation, index) => {
      requiredFields.forEach(field => {
        expect(allocation).toHaveProperty(field);
        expect(allocation[field]).toBeDefined();
        expect(allocation[field]).not.toBe('');
      });
    });
  });

  test('allocation IDs are unique', () => {
    const ids = allocationsData.map(allocation => allocation.id);
    const uniqueIds = new Set(ids);
    
    expect(uniqueIds.size).toBe(ids.length);
  });

  test('days are valid', () => {
    const validDays = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
    
    allocationsData.forEach(allocation => {
      expect(validDays).toContain(allocation.day);
    });
  });

  test('timeSlotIds follow expected format', () => {
    const timeSlotPattern = /^ts\d+$/;
    
    allocationsData.forEach(allocation => {
      expect(allocation.timeSlotId).toMatch(timeSlotPattern);
    });
  });

  test('data consistency snapshot', () => {
    // Create a summary of the data for snapshot testing
    const summary = {
      totalAllocations: allocationsData.length,
      uniqueSubjects: new Set(allocationsData.map(a => a.subjectId)).size,
      uniqueTeachers: new Set(allocationsData.filter(a => a.teacherId).map(a => a.teacherId)).size,
      uniqueRooms: new Set(allocationsData.filter(a => a.room).map(a => a.room)).size,
      uniqueTimeSlots: new Set(allocationsData.map(a => a.timeSlotId)).size,
      uniqueDays: new Set(allocationsData.map(a => a.day)).size,
      uniqueSemesters: new Set(allocationsData.map(a => a.semesterId)).size,
      
      // Count allocations by day
      dayDistribution: allocationsData.reduce((acc, allocation) => {
        acc[allocation.day] = (acc[allocation.day] || 0) + 1;
        return acc;
      }, {}),
      
      // Count allocations by semester
      semesterDistribution: allocationsData.reduce((acc, allocation) => {
        acc[allocation.semesterId] = (acc[allocation.semesterId] || 0) + 1;
        return acc;
      }, {}),
      
      // Count allocations with/without rooms
      roomAssignmentStats: {
        withRoom: allocationsData.filter(a => a.room).length,
        withoutRoom: allocationsData.filter(a => !a.room).length
      },
      
      // Count allocations with/without teachers
      teacherAssignmentStats: {
        withTeacher: allocationsData.filter(a => a.teacherId).length,
        withoutTeacher: allocationsData.filter(a => !a.teacherId).length
      }
    };
    
    expect(summary).toMatchSnapshot();
  });

  test('sort order is deterministic', () => {
    // Test that the current order can be reproduced by a deterministic sort
    const sortedAllocations = [...allocationsData].sort((a, b) => {
      // Sort by semester, then day, then timeSlot, then subject, then id
      if (a.semesterId !== b.semesterId) {
        return a.semesterId.localeCompare(b.semesterId);
      }
      
      const dayOrder = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'];
      const dayIndexA = dayOrder.indexOf(a.day);
      const dayIndexB = dayOrder.indexOf(b.day);
      if (dayIndexA !== dayIndexB) {
        return dayIndexA - dayIndexB;
      }
      
      if (a.timeSlotId !== b.timeSlotId) {
        return a.timeSlotId.localeCompare(b.timeSlotId);
      }
      
      if (a.subjectId !== b.subjectId) {
        return a.subjectId.localeCompare(b.subjectId);
      }
      
      return a.id.localeCompare(b.id);
    });
    
    // Create a simple representation for comparison
    const currentOrder = allocationsData.map(a => `${a.id}-${a.semesterId}-${a.day}-${a.timeSlotId}`);
    const expectedOrder = sortedAllocations.map(a => `${a.id}-${a.semesterId}-${a.day}-${a.timeSlotId}`);
    
    // Store the first 20 items for snapshot (to avoid huge snapshots)
    expect(currentOrder.slice(0, 20)).toMatchSnapshot();
  });

  test('field types are consistent', () => {
    const fieldTypes = {};
    
    allocationsData.forEach(allocation => {
      Object.keys(allocation).forEach(field => {
        const type = typeof allocation[field];
        const isNull = allocation[field] === null;
        const actualType = isNull ? 'null' : type;
        
        if (!fieldTypes[field]) {
          fieldTypes[field] = new Set();
        }
        fieldTypes[field].add(actualType);
      });
    });
    
    // Convert Sets to Arrays for snapshot
    const fieldTypesSummary = Object.fromEntries(
      Object.entries(fieldTypes).map(([field, types]) => [
        field,
        Array.from(types).sort()
      ])
    );
    
    expect(fieldTypesSummary).toMatchSnapshot();
  });
});
