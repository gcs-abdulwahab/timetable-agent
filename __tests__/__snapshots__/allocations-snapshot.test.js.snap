// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Allocations Data Structure allocations.json structure matches expected format 1`] = `
[
  {
    "allowedFields": [
      "day",
      "id",
      "room",
      "semesterId",
      "subjectId",
      "teacherId",
      "timeSlotId",
    ],
    "day": "string",
    "hasDay": true,
    "hasId": true,
    "hasSemesterId": true,
    "hasSubjectId": true,
    "hasTimeSlotId": true,
    "id": "string",
    "room": "string",
    "semesterId": "string",
    "subjectId": "string",
    "teacherId": "string",
    "timeSlotId": "string",
  },
  {
    "allowedFields": [
      "day",
      "id",
      "room",
      "semesterId",
      "subjectId",
      "teacherId",
      "timeSlotId",
    ],
    "day": "string",
    "hasDay": true,
    "hasId": true,
    "hasSemesterId": true,
    "hasSubjectId": true,
    "hasTimeSlotId": true,
    "id": "string",
    "room": "string",
    "semesterId": "string",
    "subjectId": "string",
    "teacherId": "string",
    "timeSlotId": "string",
  },
  {
    "allowedFields": [
      "day",
      "id",
      "room",
      "semesterId",
      "subjectId",
      "teacherId",
      "timeSlotId",
    ],
    "day": "string",
    "hasDay": true,
    "hasId": true,
    "hasSemesterId": true,
    "hasSubjectId": true,
    "hasTimeSlotId": true,
    "id": "string",
    "room": "string",
    "semesterId": "string",
    "subjectId": "string",
    "teacherId": "string",
    "timeSlotId": "string",
  },
]
`;

exports[`Allocations Data Structure data consistency snapshot 1`] = `
{
  "dayDistribution": {
    "Monday": 4,
    "Thursday": 1,
    "Tuesday": 4,
    "Wednesday": 2,
  },
  "roomAssignmentStats": {
    "withRoom": 7,
    "withoutRoom": 4,
  },
  "semesterDistribution": {
    "sem1": 7,
    "sem3": 4,
  },
  "teacherAssignmentStats": {
    "withTeacher": 11,
    "withoutTeacher": 0,
  },
  "totalAllocations": 11,
  "uniqueDays": 4,
  "uniqueRooms": 2,
  "uniqueSemesters": 2,
  "uniqueSubjects": 4,
  "uniqueTeachers": 3,
  "uniqueTimeSlots": 2,
}
`;

exports[`Allocations Data Structure field types are consistent 1`] = `
{
  "day": [
    "string",
  ],
  "id": [
    "string",
  ],
  "room": [
    "string",
  ],
  "semesterId": [
    "string",
  ],
  "subjectId": [
    "string",
  ],
  "teacherId": [
    "string",
  ],
  "timeSlotId": [
    "string",
  ],
}
`;

exports[`Allocations Data Structure sort order is deterministic 1`] = `
[
  "edited-16-sem3-Monday-ts2",
  "edited-17-sem3-Tuesday-ts2",
  "edited-18-sem3-Wednesday-ts2",
  "edited-19-sem3-Thursday-ts2",
  "edited-13-sem1-Monday-ts2",
  "edited-14-sem1-Tuesday-ts2",
  "edited-15-sem1-Wednesday-ts2",
  "edited-2-sem1-Monday-ts4",
  "edited-3-sem1-Tuesday-ts4",
  "edited-0-sem1-Monday-ts2",
  "edited-1-sem1-Tuesday-ts2",
]
`;
